File: web/src/utils/project-permissions.ts
export type ProjectRole = 'manager' | 'leader' | 'member' | 'guest';

export function projectRole(value?: string | null): ProjectRole {
  const role = String(value ?? '').toLowerCase();
  if (role === 'admin') return 'manager';
  if (role === 'viewer') return 'guest';
  return ['manager', 'leader', 'member', 'guest'].includes(role) ? role as ProjectRole : 'guest';
}

export const canViewMembers = (role?: string | null) => projectRole(role) !== 'guest';
export const canAddMembers = (role?: string | null) => ['manager', 'leader'].includes(projectRole(role));
export const canViewSettings = (role?: string | null) => projectRole(role) === 'manager';
export const canViewOtherTimesheets = (role?: string | null) => ['manager', 'leader'].includes(projectRole(role));
export const canEditOtherTimesheets = (role?: string | null) => projectRole(role) === 'manager';

export function canRemoveMember(
  callerRole: string | null | undefined,
  targetRole: string | null | undefined,
  isSelf: boolean,
  isOwner: boolean,
  targetIsOwner: boolean,
) {
  if (targetIsOwner) return false;
  if (isOwner) return true;
  const caller = projectRole(callerRole);
  const target = targetRole ? projectRole(targetRole) : null;
  if (caller === 'manager') return isSelf || target !== 'manager';
  if (caller === 'leader') return isSelf || target === null || ['member', 'guest'].includes(target);
  return false;
}

File: backend/src/domain/project_permissions.rs
//! Pure project member, settings, and timesheet permission rules.

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ProjectRole {
    Manager,
    Leader,
    Member,
    Guest,
}

impl ProjectRole {
    pub fn parse(value: &str) -> Option<Self> {
        match value.to_ascii_lowercase().as_str() {
            "manager" | "admin" => Some(Self::Manager),
            "leader" => Some(Self::Leader),
            "member" => Some(Self::Member),
            "guest" | "viewer" => Some(Self::Guest),
            _ => None,
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct ProjectActor {
    pub is_owner: bool,
    pub role: Option<ProjectRole>,
}

impl ProjectActor {
    pub fn can_view_members(self) -> bool {
        self.is_owner
            || matches!(
                self.role,
                Some(ProjectRole::Manager | ProjectRole::Leader | ProjectRole::Member)
            )
    }

    pub fn can_add_members(self) -> bool {
        self.is_owner || matches!(self.role, Some(ProjectRole::Manager | ProjectRole::Leader))
    }

    pub fn can_view_settings(self) -> bool {
        self.is_owner || self.role == Some(ProjectRole::Manager)
    }

    pub fn can_assign_role(self, role: ProjectRole) -> bool {
        self.is_owner
            || self.role == Some(ProjectRole::Manager)
            || (self.role == Some(ProjectRole::Leader)
                && matches!(role, ProjectRole::Member | ProjectRole::Guest))
    }

    pub fn can_remove(
        self,
        target_is_owner: bool,
        target_role: Option<ProjectRole>,
        is_self: bool,
    ) -> bool {
        if target_is_owner {
            return false;
        }
        if self.is_owner {
            return true;
        }
        match self.role {
            Some(ProjectRole::Manager) => is_self || target_role != Some(ProjectRole::Manager),
            Some(ProjectRole::Leader) => {
                is_self
                    || matches!(
                        target_role,
                        None | Some(ProjectRole::Member | ProjectRole::Guest)
                    )
            }
            _ => false,
        }
    }

    pub fn can_view_timesheet(self, own: bool) -> bool {
        own || self.is_owner
            || matches!(self.role, Some(ProjectRole::Manager | ProjectRole::Leader))
    }

    pub fn can_edit_timesheet(self, own: bool) -> bool {
        own || self.is_owner || self.role == Some(ProjectRole::Manager)
    }
}
